zebra-consensus/src/transaction.rs

```diff
@@ -353,24 +343,33 @@ where
             // async_checks.push(rsp);
         }
 
-        // Finally, wait for all asynchronous checks to complete
-        // successfully, or fail verification if they error.
-        while let Some(check) = async_checks.next().await {
-            tracing::trace!(?check, remaining = async_checks.len());
-            check?;
-        }
+        Self::wait_for_checks(async_checks).await?;
 
         Ok(tx.hash())
     }
 
     async fn verify_v5_transaction(
         request: Request,
         network: Network,
+        script_verifier: script::Verifier<ZS>,
+        inputs: &[transparent::Input],
     ) -> Result<transaction::Hash, TransactionError> {
-        Self::verify_v5_transaction_network_upgrade(
-            &request.transaction(),
-            request.upgrade(network),
-        )?;
+        let mut async_checks = FuturesUnordered::new();
+        let transaction = request.transaction();
+
+        Self::verify_v5_transaction_network_upgrade(&transaction, request.upgrade(network))?;
+
+        // Do basic checks first
+        check::has_inputs_and_outputs(&transaction)?;
```

Comment:
Should this be in the main `Service::call` method? Is it common to all transaction versions?

Reply:
Yes, not sure how I missed that one.

We should do it as early as possible, because other checks might assume there are inputs and outputs.

Reply:
Done.